### PR TITLE
Version finetune-lora for review

### PR DESCRIPTION
**Pull request by complexbrains**

### CODE REVIEW

File: cneuromax/projects/friends_language_encoder/peftmodule.py [hunk 1]
```diff
@@ -21,21 +20,25 @@ class PEFTLitModule(BaseLitModule, ABC):
...
+        lora_config = LoraConfig(
+            task_type=self.peft_config.task_type,
+            lora_alpha=self.peft_config.lora_alpha,
+            r=self.peft_config.lora_rank,
+            lora_dropout=self.peft_config.lora_dropout,
         )
-        self.nnmodule = get_peft_model(nnmodule, peft_config)
-        self.optimizer_partial = optimizer
-        self.scheduler_partial = scheduler
+        self.nnmodule = get_peft_model(self.nnmodule, lora_config)
```
Comment (maximilienleclei) [inline cneuromax/projects/friends_language_encoder/peftmodule.py]: you don't need to reinstantiate a LoraConfig object. LoraConfig is a PeftConfig. If this doesn't make complete sense, brush up on python inheritance

Comment (maximilienleclei) [reply cneuromax/projects/friends_language_encoder/peftmodule.py]: all of this can be replaced by `self.nnmodule = get_peft_model(self.nnmodule, peft_config)`

File: cneuromax/projects/friends_language_encoder/peftmodule.py [hunk 2]
```diff
@@ -21,21 +20,25 @@ class PEFTLitModule(BaseLitModule, ABC):
 
     def __init__(
         self: "PEFTLitModule",
-        config: BaseLitModuleConfig,
-        nnmodule: nn.Module,
-        optimizer: partial[Optimizer],
-        scheduler: partial[LRScheduler],
+        config: "BaseLitModuleConfig",
+        peft_config: "PeftConfig",
```
Comment (maximilienleclei) [inline cneuromax/projects/friends_language_encoder/peftmodule.py]: doesn't need to be in parenthesis

File: cneuromax/projects/friends_language_encoder/peftmodule.py [hunk 3]
```diff
@@ -21,21 +20,25 @@ class PEFTLitModule(BaseLitModule, ABC):
...
     ) -> None:
         """."""
-        super().__init__(config, nnmodule, optimizer, scheduler)
+        super().__init__(
+            config,
+            peft_config,
+            *args,
+            **kwargs,
+        )
         self.config = config
```
Comment (maximilienleclei) [inline cneuromax/projects/friends_language_encoder/peftmodule.py]: since you're not doing anything with the `config` object, get rid of it, the `args` `kwargs` will take care of it

File: cneuromax/projects/friends_language_encoder/peftmodule.py [hunk 4]
```diff
@@ -21,21 +20,25 @@ class PEFTLitModule(BaseLitModule, ABC):
...
+        config: "BaseLitModuleConfig",
+        peft_config: "PeftConfig",
+        *args: Any,  # noqa: ANN401, ARG002
+        **kwargs: Any,  # noqa: ANN401, ARG002
     ) -> None:
         """."""
-        super().__init__(config, nnmodule, optimizer, scheduler)
+        super().__init__(
+            config,
+            peft_config,
```
Comment (maximilienleclei) [inline cneuromax/projects/friends_language_encoder/peftmodule.py]: this is wrong, `BaseLitModule`'s constructor does not admita  `peft_config` argument

File: cneuromax/projects/friends_language_encoder/litmodule_peft.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    """."""
+
+    def __init__(
+        self: "FriendsPEFTModule",
+        peft_config: PeftConfig,
+        *args: Any,  # noqa: ANN401, ARG002
+        **kwargs: Any,  # noqa: ANN401, ARG002
+    ) -> None:
+        super().__init__(peft_config, *args, **kwargs)
+        self.nnmodule = get_peft_model(self.nnmodiule, peft_config)
```
Comment (maximilienleclei) [inline cneuromax/projects/friends_language_encoder/litmodule_peft.py]: guessing this is still a draft. you dont need to implement this constructor, it is done by default, if my comment doesn't make sense brush up on method inheritance

File: cneuromax/projects/friends_language_encoder/litmodule_peft.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                otherwise the logits.
+        """
+        out = self.nnmodule(
+            input_ids=batch["input_ids"],
+            attention_mask=batch["attention_mask"],
+            labels=batch["labels"],
+        )
+        out: Tensor = (
+            out["loss"] if stage in ["train", "val", "test"] else out["logits"]
+        )
```
Comment (maximilienleclei) [inline cneuromax/projects/friends_language_encoder/litmodule_peft.py]: `step` method always returns the loss. If you want the `stage == perdict` to return the logits, implement the `predict_step` method.

File: cneuromax/projects/friends_language_encoder/litmodule_peft.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+@dataclass
+class FriendsLitModuleConfig(BaseLitModuleConfig):
+    """Holds :class:`FriendsLitModule` config values.
+
+    Args:
+        layer_name: layer to unfreeze
+    """
+
+    layer_names: str = "${layer_names}"
```
Comment (maximilienleclei) [inline cneuromax/projects/friends_language_encoder/litmodule_peft.py]: 1. are you sure this is a string? looks like it's a list from your docstring
2. donc set the config to `${layer_names}` if you need a default set it to `field(default_factory=list)` if it's a list of string, or `""` if it's a string

**Review by maximilienleclei — COMMENTED** [review]
.

File: cneuromax/projects/friends_language_encoder/litmodule_peft.py (hunk 7 above)
Comment (complexbrains) [reply cneuromax/projects/friends_language_encoder/litmodule_peft.py]: it  is `layer_names: list[Any, Any] = "${layer_names}"`

File: cneuromax/projects/friends_language_encoder/peftmodule.py (hunk 2 above)
Comment (complexbrains) [reply cneuromax/projects/friends_language_encoder/peftmodule.py]: do you mean like this? 
```
@dataclass
class PEFTLitModule(BaseLitModule, ABC):
    """``project`` :class:`~PEFTLitModule`."""

    def __init__(
        self: "PEFTLitModule",
        *args: Any,  # noqa: ANN401, ARG002
        **kwargs: Any,  # noqa: ANN401, ARG002
    ) -> None:
        """."""
        super().__init__(
            *args,
            **kwargs,
        )
        peft_config = PeftConfig()
        self.nnmodule = get_peft_model(self.nnmodule, peft_config)
```

File: cneuromax/projects/friends_language_encoder/litmodule_peft.py (hunk 5 above)
Comment (complexbrains) [reply cneuromax/projects/friends_language_encoder/litmodule_peft.py]: Yes sorry, but I understand. I guess my confusion comes from we already also have 
` self.nnmodule = get_peft_model(self.nnmodiule, peft_config)`  in `PEFTLitModule` as well. So then calling it here also makes no sense, unless we pass the hydra config to `get_peft_model` where my parameters are in.. 

````
class FriendsPEFTModule(PEFTLitModule):
    """."""

    def __init__(
        self: "FriendsPEFTModule",
        *args: Any,  # noqa: ANN401, ARG002
        **kwargs: Any,  # noqa: ANN401, ARG002
    ) -> None:
        super().__init__(*args, **kwargs)
        self.nnmodule = get_peft_model(self.nnmodiule, self.config)
````

otherwise where we gonna read the hydra parameters from?

File: cneuromax/projects/friends_language_encoder/peftmodule.py (hunk 2 above)
Comment (maximilienleclei) [reply cneuromax/projects/friends_language_encoder/peftmodule.py]: my bad quotations not parenthesis, meant that `peft_config: "PeftConfig"` can turn to `peft_config: PeftConfig`

File: cneuromax/projects/friends_language_encoder/litmodule_peft.py (hunk 6 above)
Comment (complexbrains) [reply cneuromax/projects/friends_language_encoder/litmodule_peft.py]: I guess this was how you suggested/implemented back then we had this confusion with the whether to  output logits or loss.. so then you went with loss.. but will check this..

File: cneuromax/projects/friends_language_encoder/litmodule_peft.py (hunk 5 above)
Comment (maximilienleclei) [reply cneuromax/projects/friends_language_encoder/litmodule_peft.py]:
```
class FriendsFinetuningModel(PEFTLitModule):
    """``project`` :class:`~BaseLitModule`."""

    def __init__(
        self: "FriendsFinetuningModel",
        *args: Any,  # noqa: ANN401, ARG002
        **kwargs: Any,  # noqa: ANN401, ARG002
    ) -> None:
        super().__init__(*args, **kwargs)
        self.config: FriendsLitModuleConfig

        for param in self.nnmodule.parameters():
            param.requires_grad = False
        ...
```
will work

File: cneuromax/projects/friends_language_encoder/litmodule_peft.py (hunk 7 above)
Comment (maximilienleclei) [reply cneuromax/projects/friends_language_encoder/litmodule_peft.py]: alright then smth like `layer_names: list[str] = field(default_factory=list)` should do (means the default `[]` empty list). 
See https://docs.python.org/3/library/dataclasses.html#dataclasses.field

File: cneuromax/projects/friends_language_encoder/peftmodule.py (hunk 2 above)
Comment (complexbrains) [reply cneuromax/projects/friends_language_encoder/peftmodule.py]: ah ok makes sense, otherwise it complains for sure!

File: cneuromax/projects/friends_language_encoder/litmodule_peft.py (hunk 6 above)
Comment (maximilienleclei) [reply cneuromax/projects/friends_language_encoder/litmodule_peft.py]: yeah `step` is supposed to output a loss. Use `predict_step` https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.core.LightningModule.html#lightning.pytorch.core.LightningModule.predict_step for inference

File: cneuromax/projects/friends_language_encoder/litmodule_peft.py [hunk 8]
```diff
@@ -41,18 +37,9 @@ def __init__(
...
-
-        for layer_name in self.config.layer_names:
-            for name, param in self.nnmodule.named_parameters():
-                if layer_name in name:
-                    param.requires_grad = True
-
     def step(
         self: "FriendsFinetuningModel",
         batch: BatchEncoding,
-        stage: Any[str, one_of("train", "val", "test", "predict")],
```
Comment (maximilienleclei) [inline cneuromax/projects/friends_language_encoder/litmodule_peft.py]: this has to stay, see https://github.com/courtois-neuromod/cneuromax/blob/main/cneuromax/fitting/deeplearning/litmodule/base.py#L50

**Review by maximilienleclei — COMMENTED** [review]
looks good
